sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSink.java

```diff
@@ -0,0 +1,335 @@
+      long timestampMsSinceEpoch = BigEndianLongCoder.of().decode(inStream, Context.NESTED);
+      return new PubsubClient.OutgoingMessage(elementBytes, timestampMsSinceEpoch);
+    }
+  }
+
+  @VisibleForTesting
+  static final Coder<PubsubClient.OutgoingMessage> CODER = new OutgoingMessageCoder();
+
+  // ================================================================================
+  // ShardFv
+  // ================================================================================
+
+  /**
+   * Convert elements to messages and shard them.
+   */
+  private static class ShardFn<T> extends DoFn<T, KV<Integer, PubsubClient.OutgoingMessage>> {
+    private final Aggregator<Long, Long> elementCounter =
+        createAggregator("elements", new Sum.SumLongFn());
+    private final Coder<T> elementCoder;
+    private final int numShards;
+
+    ShardFn(Coder<T> elementCoder, int numShards) {
+      this.elementCoder = elementCoder;
+      this.numShards = numShards;
+    }
+
+    @Override
+    public void processElement(ProcessContext c) throws Exception {
+      elementCounter.addValue(1L);
+      byte[] elementBytes = CoderUtils.encodeToByteArray(elementCoder, c.element());
+      long timestampMsSinceEpoch = c.timestamp().getMillis();
+      c.output(KV.of(ThreadLocalRandom.current().nextInt(numShards),
+                     new PubsubClient.OutgoingMessage(elementBytes, timestampMsSinceEpoch)));
+    }
+  }
+
+  // ================================================================================
+  // WriterFn
+  // ================================================================================
+
+  /**
+   * Publish messages to Pubsub in batches.
+   */
+  private static class WriterFn
+      extends DoFn<KV<Integer, Iterable<PubsubClient.OutgoingMessage>>, Void> {
+    private final PubsubClientFactory pubsubFactory;
+    private final PubsubClient.TopicPath topic;
+    private final String timestampLabel;
+    private final String idLabel;
+
+    /**
+     * Client on which to talk to Pubsub. Null until created by {@link #startBundle}.
+     */
+    @Nullable
+    private transient PubsubClient pubsubClient;
+
+    private final Aggregator<Long, Long> batchCounter =
+        createAggregator("batches", new Sum.SumLongFn());
+    private final Aggregator<Long, Long> elementCounter =
+        createAggregator("elements", new Sum.SumLongFn());
+    private final Aggregator<Long, Long> byteCounter =
+        createAggregator("bytes", new Sum.SumLongFn());
+
+    WriterFn(PubsubClientFactory pubsubFactory, PubsubClient.TopicPath topic,
+             String timestampLabel, String idLabel) {
+      this.pubsubFactory = pubsubFactory;
+      this.topic = topic;
+      this.timestampLabel = timestampLabel;
+      this.idLabel = idLabel;
+    }
+
+    /**
+     * BLOCKING
+     * Send {@code messages} as a batch to Pubsub.
+     */
+    private void publishBatch(List<PubsubClient.OutgoingMessage> messages, int bytes)
```

Comment:
I don't much care, but you write this as both `OutgoingMessage` and `PubsubClient.OutgoingMessage` in this file. Probably could pick one and stick with it.